src/Features/CSharp/Portable/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs

```diff
@@ -86,5 +108,139 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
+                    //
+                    // ---
+                    // INTENTIONAL FALLTHROUGH
+                }
+                else
+                {
+                    var firstTokenAfterEmbeddedStatement = embeddedStatement.GetLastToken().GetNextToken();
+                    if (!FormattingRangeHelper.AreTwoTokensOnSameLine(firstTokenAfterEmbeddedStatement, statement.GetLastToken()))
+                    {
+                        // The part of the statement following the embedded statement does not fit on one line. Examples:
+                        //
+                        //   do
+                        //     SomeMethod();
+                        //   while (x < 0 ||    // <-- This condition is split across multiple lines.
+                        //          x > 10);
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private static bool RequiresBracesToMatchContext(SyntaxNode statement, StatementSyntax embeddedStatement)
+        {
+            if (!statement.IsKind(SyntaxKind.IfStatement, SyntaxKind.ElseClause))
+            {
+                // 'if' statements are the only statements that can have multiple embedded statements which are
+                // considered relative to each other.
+                return false;
+            }
+
+            var topLevelContext = GetTopLevelContext(statement);
+            if (AnyPartUsesBraces(topLevelContext))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static IfStatementSyntax GetTopLevelContext(SyntaxNode statement)
+        {
+            IfStatementSyntax result;
+            if (statement.IsKind(SyntaxKind.ElseClause))
+            {
+                result = (IfStatementSyntax)statement.Parent;
+            }
+            else
+            {
+                Debug.Assert(statement.IsKind(SyntaxKind.IfStatement));
+                result = (IfStatementSyntax)statement;
+            }
+
+            while (result != null)
+            {
+                if (!result.IsParentKind(SyntaxKind.ElseClause))
+                {
+                    break;
+                }
+
+                result = (IfStatementSyntax)result.Parent.Parent;
+            }
+
+            return result;
+        }
+
+        private static bool AnyPartUsesBraces(IfStatementSyntax statement)
+        {
+            // Iterative instead of recursive to avoid stack depth problems
+            while (statement != null)
+            {
+                if (statement.Statement.IsKind(SyntaxKind.Block))
+                {
+                    return true;
+                }
+
+                var elseStatement = statement.Else?.Statement;
+                if (elseStatement.IsKind(SyntaxKind.Block))
+                {
+                    return true;
+                }
+
+                statement = elseStatement as IfStatementSyntax;
```

Comment:
i don't quite understand this.  above, you check both the true/false parts for braces (that i get), but here you only walk into the else part?  

This seems correct in terms of checking a chain of `if ... else if ... else if ... else`.  However, it's unclear to me why you don't need to look int the true-part as well to see if that has blocks nested in it.

Reply:
note: i can see that this is the semantics you're going for.  would be good to comment this explaining as such.

Reply:
> However, it's unclear to me why you don't need to look int the true-part as well to see if that has blocks nested in it.

Adding a test for this, and documenting the behavior where the test is defined.

Reply:
➡️ This should now be resolved